### PR TITLE
test(esl-mixin): additional prototype tests

### PR DESCRIPTION
**Pull request by fshovchko**
Relates to: #1786

### CODE REVIEW

**Review by dshovchko — APPROVED** [review]

File: src/modules/esl-mixin-element/test/mixin.test.ts [hunk 1]
```diff
@@ -101,4 +101,108 @@ describe('ESLMixinElement', () => {
...
+      const attrName = 'test-attr';
+
+      beforeAll(() => $el = CTestMixin.get($host) as CTestMixin);
+
+      test('should return the attribute value', () => {
+        const val = $host.getAttribute(attrName);
+        expect($el.$$attr(attrName)).toBe(val);
+      });
+
+      test('should set the attribute value and return the new value', () => {
```
Comment (ala-n) [inline src/modules/esl-mixin-element/test/mixin.test.ts:127]:
```suggestion
      test('should set the attribute value and return the previous value', () => {
```

File: src/modules/esl-mixin-element/test/mixin.test.ts [hunk 2]
```diff
@@ -101,4 +101,108 @@ describe('ESLMixinElement', () => {
...
+    describe('ESLMixinElement $attr - boolean', () => {
+      let el: CTestMixin;
+      const attrName = 'test-attr-bool';
+
+      beforeAll(() => el = CTestMixin.get($host) as CTestMixin);
+
+      test('should return the initial attribute value', () => {
+        const val = $host.getAttribute(attrName);
+        expect(el.$$attr(attrName)).toBe(val);
+      });
```
Comment (ala-n) [inline src/modules/esl-mixin-element/test/mixin.test.ts:142]: Duplicates test on lines 122-124

File: src/modules/esl-mixin-element/test/mixin.test.ts [hunk 3]
```diff
@@ -101,4 +101,108 @@ describe('ESLMixinElement', () => {
...
+
+      test('should set and return an empty string for boolean true', () => {
+        expect(el.$$attr(attrName, true)).toBe(null);
+        expect(el.$$attr(attrName)).toBe('');
+      });
+
+      test('should set and return null for removed attribute', () => {
+        expect(el.$$attr(attrName, false)).toBe('');
+        expect(el.$$attr(attrName)).toBe(null);
+      });
```
Comment (ala-n) [inline src/modules/esl-mixin-element/test/mixin.test.ts:152]: Consider merge this test group with the `ESLMixinElement $attr`

File: src/modules/esl-mixin-element/test/mixin.test.ts [hunk 4]
```diff
@@ -101,4 +101,108 @@ describe('ESLMixinElement', () => {
...
+    $host.toggleAttribute(CTestMixin.is, true);
+    CTestMixin.register();
+
+    beforeAll(() => document.body.appendChild($host));
+
+    describe('ESLMixinElement $attr', () => {
+      let $el: CTestMixin;
+      const attrName = 'test-attr';
+
+      beforeAll(() => $el = CTestMixin.get($host) as CTestMixin);
```
Comment (ala-n) [inline src/modules/esl-mixin-element/test/mixin.test.ts:120]: Consider moving and merging with beforeAll on the top level to get rid of repetitive code
To deal with async nature of registration.
Register mixin after adding it to the dom or make beforeAll step async with a microtask waiter trick:
`await Promise.resolve()`

File: src/modules/esl-mixin-element/test/mixin.test.ts [hunk 5]
```diff
@@ -101,4 +101,108 @@ describe('ESLMixinElement', () => {
...
+      });
+    });
+
+    test('ESLMixinElement $$fire', (done) => {
+      const $el = CTestMixin.get($host) as CTestMixin;
+
+      $host.addEventListener('testevent', (e) => {
+        expect(e).toBeInstanceOf(CustomEvent);
+        done();
+      }, {once: true});
```
Comment (ala-n) [inline src/modules/esl-mixin-element/test/mixin.test.ts:194]: Cosmetic: mock $host.dispatchEvent instead

File: src/modules/esl-mixin-element/test/mixin.test.ts [hunk 6]
```diff
@@ -101,4 +101,108 @@ describe('ESLMixinElement', () => {
...
+        $el.$$cls('a', true);
+        $el.$$cls('b', true);
+        expect($host.className).toBe('a b');
+      });
+
+      test('should remove the class when setting to false', () => {
+        $host.className = 'a b';
+        expect($el.$$cls('a b', false)).toBe(false);
+        expect($host.className).toBe('');
+      });
```
Comment (ala-n) [inline src/modules/esl-mixin-element/test/mixin.test.ts:185]: Cosmetic: it looks ok to merge it with the get cases

**Review by julia-murashko — APPROVED** [review]

**Review by yadamskaya — APPROVED** [review]

File: src/modules/esl-mixin-element/test/mixin.test.ts [hunk 7]
```diff
@@ -101,4 +101,83 @@ describe('ESLMixinElement', () => {
...
+    });
+
+    test('ESLMixinElement $$fire', () => {
+      const eventName = 'testevent';
+      $host.dispatchEvent = jest.fn();
+      $el.$$fire(eventName);
+
+      const event: CustomEvent = ($host.dispatchEvent as jest.Mock).mock.calls[0][0];
```
Comment (ala-n) [inline src/modules/esl-mixin-element/test/mixin.test.ts:176]:
```
expect(mock).lastCalledWith(expect.objectContaining({...}))
```
https://jestjs.io/docs/expect#expectobjectcontainingobject

**Review by ala-n — APPROVED** [review]

**Pull request merged**